src/Autofac/Core/Resolving/InstanceLookup.cs

```diff
@@ -83,13 +84,21 @@ public object Execute()
             _executed = true;
 
             var sharing = _decoratorTargetComponent?.Sharing ?? ComponentRegistration.Sharing;
+            var sharedId = ComponentRegistration.Id;
+            if (_decoratorTargetComponent != null)
+            {
+                // Issue #1113
+                // Need to generate a GUID that represents the composite of the decorator and target registrations
+                // to retain multiple shared instances for IEnumerable<T>
+                sharedId = DeterministicGuid.Create(_decoratorTargetComponent.Id, sharedId.ToByteArray());
```

Comment:
Since it is possible for multiple instances of a given decorator to be generated via IEnumerable<T>, the ID used to store said instances can't all be the ID for the decorator registration, since that just stomps all over itself. "Namespacing" the decorator registration ID with the registration ID of the target creates uniqueness with a very low chance of collision.